### ISSUE TITLE
On upgrade to 0.7.0 all calls to engine.execute() are hanging

### ISSUE THREAD

**Issue by djfroofy** [alex/alchimia#30]
Python 2.7.5
Twisted==16.4.1
alchimia==0.7.0
psycopg2==2.6.2
SQLAlchemy==1.0.15
Linux ahost-02 3.10.0-229.7.2.el7.x86_64 #1 SMP Tue Jun 23 05:02:16 PDT 2015 x86_64 x86_64 x86_64 GNU/Linux

In my application calling engine.execute() results in hang - Deferred is never called. I injected verbose logging in alchimia/engine.py and it appears the inner function "container" in _defer_to_worker is not being called:

```
2016-10-05T05:58:43Z [DEBUG] a.e.TwistedEngine - execute called. args=(<sqlalchemy.sql.selectable.Select at 0x4b267d0; Select object>,) kwargs={}
2016-10-05T05:58:43Z [DEBUG] a.e.TwistedEngine - _defer_to_engine called. a=(<sqlalchemy.sql.selectable.Select at 0x4b267d0; Select object>,) f=<bound method Engine.execute of Engine(postgresql+psycopg2://zzz:***@localhost/zzz)> k={}
2016-10-05T05:58:43Z [DEBUG] a.engine - _defer_to_worker called. args=(<sqlalchemy.sql.selectable.Select at 0x4b267d0; Select object>,) worker=<twisted._threads._threadworker.ThreadWorker object at 0x499e290> work=<bound method Engine.execute of Engine(postgresql+psycopg2://zzz:***@localhost/zzz)> deliver=<bound method EPollReactor.callFromThread of <twisted.internet.epollreactor.EPollReactor object at 0x3224fd0>> kwargs={}
```

If I change related code to call to engine.connect() the call succeeds as usual. Injected logging:

```
2016-10-05T06:16:30Z [DEBUG] d.c.FooController - listing foo
2016-10-05T06:16:30Z [DEBUG] a.e.TwistedEngine - connect: called.
2016-10-05T06:16:30Z [DEBUG] a.engine - _threaded_worker#_start_worker: starting thread. target=<function work at 0x50edb18>
2016-10-05T06:16:30Z [DEBUG] a.engine - _threaded_worker#_start_worker: daemonized thead ... starting. thread=<Thread(Thread-3, initial daemon)>
2016-10-05T06:16:30Z [DEBUG] a.e.TwistedEngine - connect: deferring to worker. worker=<twisted._threads._threadworker.ThreadWorker object at 0x50e4f10>
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker called. args=() worker=<twisted._threads._threadworker.ThreadWorker object at 0x50e4f10> work=<bound method Engine.connect of Engine(postgresql+psycopg2://zzz:***@localhost/zzz)> deliver=<bound method EPollReactor.callFromThread of <twisted.internet.epollreactor.EPollReactor object at 0x33293d0>> kwargs={}
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: running.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: got result. result=<sqlalchemy.engine.base.Connection object at 0x50f67d0>
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: delivered result.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker called. args=() worker=<twisted._threads._threadworker.ThreadWorker object at 0x50e4f10> work=<bound method Connection.begin of <sqlalchemy.engine.base.Connection object at 0x50f67d0>> deliver=<bound method EPollReactor.callFromThread of <twisted.internet.epollreactor.EPollReactor object at 0x33293d0>> kwargs={}
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: running.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: got result. result=<sqlalchemy.engine.base.RootTransaction object at 0x50e4b50>
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: delivered result.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker called. args=(<sqlalchemy.sql.selectable.Select at 0x50f6810; Select object>,) worker=<twisted._threads._threadworker.ThreadWorker object at 0x50e4f10> work=<bound method Connection.execute of <sqlalchemy.engine.base.Connection object at 0x50f67d0>> deliver=<bound method EPollReactor.callFromThread of <twisted.internet.epollreactor.EPollReactor object at 0x33293d0>> kwargs={}
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: running.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: got result. result=<sqlalchemy.engine.result.ResultProxy object at 0x50f64d0>
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: delivered result.
2016-10-05T06:16:30Z [DEBUG] d.datastore - got result. result=<alchimia.engine.TwistedResultProxy object at 0x50f68d0>
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker called. args=() worker=<twisted._threads._threadworker.ThreadWorker object at 0x50e4f10> work=<bound method ResultProxy.fetchall of <sqlalchemy.engine.result.ResultProxy object at 0x50f64d0>> deliver=<bound method EPollReactor.callFromThread of <twisted.internet.epollreactor.EPollReactor object at 0x33293d0>> kwargs={}
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: running.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: got result. result=[…<snip>…]
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: delivered result.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker called. args=() worker=<twisted._threads._threadworker.ThreadWorker object at 0x50e4f10> work=<bound method RootTransaction.commit of <sqlalchemy.engine.base.RootTransaction object at 0x50e4b50>> deliver=<bound method EPollReactor.callFromThread of <twisted.internet.epollreactor.EPollReactor object at 0x33293d0>> kwargs={}
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: running.
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: got result. result=None
2016-10-05T06:16:30Z [DEBUG] a.engine - _defer_to_worker#container: delivered result.
2016-10-05T06:16:30Z [INFO] t.p.log - "127.0.0.1" - - [05/Oct/2016:06:16:29 +0000] "GET /foo HTTP/1.1" 200 1478 "-" "-".
```

The code in question that hangs:

```
     def get_foo(self):
        foo = tables.foo
        stmt = select([foo.c.foo_id, foo.c.modified])
        stmt = stmt.order_by(foo.c.foo_id.asc())
        d = self.engine.execute(stmt)
        return d.addCallback(blah) # blah just prints result ... but this is never called back
```

**Comment by johnwlockwood** [comment]
I to experienced my tests not passing. Hanging and then failing with pool exhaustion.

**Comment by glyph** [comment]
So, from what I can tell, Alchimia works as designed, but, there is perhaps a problem that can be addressed here.

Here's a little stress test I wrote:

```python3
from twisted.logger import globalLogBeginner, textFileLogObserver
import traceback
from alchimia import wrap_engine
from sqlalchemy import create_engine
from twisted.internet import reactor
from twisted.internet.defer import ensureDeferred
from twisted.internet.task import LoopingCall

engine = wrap_engine(
    reactor, create_engine('postgresql://postgres@127.0.0.1:6543/template1')
)

all_connections = []

async def connectify():
    print("connecting...", len(all_connections))
    try:
        connection = (await engine.connect())
    except:
        traceback.print_exc()
        return
    all_connections.append(connection)
    print("connected! total connections:", len(all_connections))
    for x in range(100):
        await connection.execute("select pg_sleep(0.1)")
        print("useful_work_unit", x)
    print("closing...")
    all_connections.remove(connection)
    await connection.close()
    print("closed!")
print("constructing")
def doed():
    ensureDeferred(connectify())
lc = LoopingCall(doed)
print("starting")
lc.start(0.231)
print("running")
import sys
globalLogBeginner.beginLoggingTo([textFileLogObserver(sys.stdout)])
reactor.run()
```

**Comment by glyph** [comment]
Running this, I do eventually see the traceback you all are talking about.  However, it's only once the pool is legitimately exhausted.

However, If I change this to delete the `await connection.close()` line, instead connections very rapidly start to pile up, and no useful work gets done.

This problem is considerably worse in Alchimia than it is in normal SQLAlchemy because even if you lose track of your `connection` object completely, there's a worker thread open in the background that will happily keep it open and blocking forever.

One thing that occurs to me is to turn the "weakness" of async here into a strength, and set a deadline on every checked-out connection, invalidating it if you don't put it back in the pool after too long.

**Comment by glyph** [comment]
Without a clear reproducer, I'm going to close this, since I'm assuming the OP has long since moved on.  I'll file some more issues for clearer *possible* causes, and if you have an actual reproducer that demonstrates the bug, please open another one with it.
